### ISSUE TITLE
TreeView: add a onClick prop when clicking on nodes

### ISSUE THREAD

**Issue by julien** [liferay/clay#4568]
TreeView has a "onSelectionChange" prop, but that only works
when the component uses a checkbox.

In DXP though, some Trees don't have checkboxes but are selectable, 
for example:

![](https://user-images.githubusercontent.com/5572/149171005-9a25a403-b84b-41eb-b8d6-cd39d98cefe8.png)

Currently, clicking on one of these nodes marks it as selected.
So we need to think about that an implement it.

**Comment by julien** [comment]
This is actually not needed, the `TreeViewItem` component already can handle this by having an `onClick` prop. Sorry for the noise

**Comment by julien** [comment]
Re-opening, because clicking on our tree "expands" the tree, but we need to be able to expand and select.

**Comment by julien** [comment]
@matuzalemsteles to give you a bit more context, in a lot of use cases (DXP), when we click on a node we should actually "select" it ... I'm still figuring out what the best way of doing this would be, because clicking on nodes that have children, expands them and we'll need to "expand + select" which isn't very easy to think about. Maybe we can settle on using double click? I'm not sure - but let's talk about this when we can.

**Comment by matuzalemsteles** [comment]
@julien Hmm, yeah that's pretty weird, doesn't it show any visual effect that elements are selected? I don't know if adding an `onClick` can help because it would have to deal with the whole multiple selection mechanisms, if it's a single selection it might not be a problem.

**Comment by julien** [comment]
What is pretty weird?

> Doesn't it show any visual effect that elements are selected?
No, but I can improve that.

>  I don't know if adding an onClick can help because it would have to deal with the whole multiple selection mechanisms, if it's a single selection it might not be a problem.

I added this to solve most use cases in DXP, but you're right, we'll need to make the multiple selection work as well. Until this is fixed, our TreeView can't be used in DXP.

**Comment by matuzalemsteles** [comment]
I think we can add the methods that trigger the selection via render props so we can allow anything to trigger a selection.

Something like that:
```jsx
<TreeView>
  {(item, selection) => (
    <TreeView.Item className="parent-list-item">
      <TreeView.ItemStack onClick={() => selection.toggleSelection(item.key)}>
        <Icon symbol="folder" />
        {item.name}
      </TreeView.ItemStack>
      <TreeView.Group items={item.children}>
        {(item) => (
          <TreeView.Item className="nested-list-item" onClick={() => selection.toggleSelection(item.key)}>
            {item.name}
          </TreeView.Item>
        )}
      </TreeView.Group>
    </TreeView.Item>
  )}
</TreeView>
```

**Comment by julien** [comment]
@matuzalemsteles ... no sure what "selection" is supposed to be in this case?

**Comment by julien** [comment]
@matuzalemsteles I don't think it makes sense to expose the `selection` that we use internally, because when the `onSelectedChange` callback is invoked, it receives a `key` argument which is a `Set` of "keys" for the selected items - which doesn't help the developer at all. What we want to do, is let the developer know which "nodes" are selected, not just their "key".

**Comment by matuzalemsteles** [comment]
@julien yes, exposing the `selection` is not very cool but we can just the methods some methods of the selection that the developer will need, in that case we can just expose access to the `toggleSelection` and create a method to get the selected key instead of exposing access straight to `selectedKeys` that the developer would do more stuff that wouldn't work, so it would just be a readonly property.

```ts
type Selection = {
  toggleSelection: (key: string) => void
  hasSelect: (key: string) => boolean
};
```

Why am I saying this?! as the case is that the developer has to make a custom selection visually speaking, he needs to trigger the selection of the element, remember that the `key` used can be equal to the `id` of the data or the `key` that is passed to the element of the `TreeView.Item`, the developer knows what the `key` is to select elements. Remember that keys can be controlled, you just need to declare something like:

```tsx
const [selectedKeys, setSelectionChange] = useState(
  new Set(['Foo', 'Bar', 'Baz'])
);

return (
  <TreeView
     onSelectionChange={setSelectionChange}
     selectedKeys={selectedKeys}
  >
    {...}
  </TreeView>
);
```

A pseudocode again showing all of this would look like this:

```tsx
const [selectedKeys, setSelectionChange] = useState(
  new Set(['Foo', 'Bar', 'Baz'])
);

return (
  <TreeView
     onSelectionChange={setSelectionChange}
     selectedKeys={selectedKeys}
  >
    {(item, selection) => (
      <TreeView.Item className="parent-list-item">
        <TreeView.ItemStack onClick={() => selection.toggleSelection(item.id)}>
          <Icon symbol="folder" />
          {item.name}
        </TreeView.ItemStack>
        <TreeView.Group items={item.children}>
          {(item) => (
            <TreeView.Item className="nested-list-item" onClick={() => selection.toggleSelection(item.id)}>
              {item.name}
            </TreeView.Item>
          )}
        </TreeView.Group>
      </TreeView.Item>
    )}
  </TreeView>
);
```

Let me know what you think, I can send you a PR to demonstrate this usage. I think this would have more complete multi-selection support because we have a more complete built-in implementation that doesn't need to traverse the tree to look for the selected elements or add a property that it's selected.

**Comment by matuzalemsteles** [comment]
@julien I sent a PR to demonstrate the use of this implementation in #4603, also with some fixes and add the behavior of selecting a single node in the entire TreeView, I added more information about this in issue #4602. Let me know what you think.

**Comment by julien** [comment]
Fixed by #4603 - sorry for going in circle, thanks @matuzalemsteles
